provider/awssd/aws_sd.go

```diff
@@ -264,6 +419,76 @@ func (p *AWSSDProvider) updatesToCreates(changes *plan.Changes) (creates []*endp
 	return creates, deletes
 }
 
+// DedupDeletesAndCreates removes targets that appear identically as both deletes and creates.
+// These targets are redundant and result in overlapping API calls. Without deduplication, RegisterInstance could be
+// invoked while DeregisterInstance is still in progress in AWS, resulting in failure to register the instance, and
+// therefore in service disruption. Redundant targets may have been introduced by updatesToCreates or srvChangesHostnameToIP.
+func (p *AWSSDProvider) DedupDeletesAndCreates(creates []*endpoint.Endpoint, deletes []*endpoint.Endpoint) ([]*endpoint.Endpoint, []*endpoint.Endpoint, error) {
```

Comment:
I share the concern that Cloud Map cannot gracefully handle `Register` and `Deregister` for the same instance at the same time. But I am not sure how can this even happen. Why would the planner generate such a plan? Is the `update` operation a concern? I would skip it for now and fix the `update` story in a separate PR.

Reply:
Thanks a lot for your comment.

> Why would the planner generate such a plan? Is the update operation a concern?

Yes, the update operation is the concern (see the example below to replicate the issue). Duplicate targets may be generated either by `updatesToCreates` or `srvChangesHostnameToIP`. The proposed `DedupDeletesAndCreates` solves pre-existing issues with A records, but also prevent issues with proposed SRV records. Since it impacts SRV records, I would be in favor of keeping `DedupDeletesAndCreates` in this PR, in order to merge an efficient handler of SRV records from the beginning.

Example: 

1. At the beginning, cluster has 1 node: 10.0.0.1. Service nginx uses nodePort 30080. CloudMap is up-to-date on A instances, so `Records` returns the following A endpoint: `nginx.abc.local 300 IN A  10.0.0.1`.
2. A new node 10.0.0.2 is added. The planner generates a change to update the A endpoint to `nginx.abc.local 300 IN A  10.0.0.1;10.0.0.2`.
3. Planned change for A record is handled by `ApplyChanges`. An `update` change is received, where `UpdateOld` is `nginx.abc.local 300 IN A  10.0.0.1` and `UpdateNew` is `nginx.abc.local 300 IN A  10.0.0.1;10.0.0.2`. `updatesToCreates` then converts the update object into three actions: `Delete` of target `10.0.0.1` and `Create` of targets `10.0.0.1` and `10.0.0.2`.

The three actions generated at point 3 are the concern. The current implementation de-registers and suddenly re-registers `10.0.0.1` instance in Cloud Map. Often, it results in a fatal 400 error `DuplicateRequest: Another operation of type DeregisterInstance and id XXXXXXX-XXX is in progress`. The implementation is idempotent, so over cycles the expected status would eventually be reached. But in case of many nodes/services or in case of frequent updates, the system may not have enough time to reach the expected status.

The proposed implementation fixes the issue by simply removing all `Delete` and `Create` operations that are identical (i.e. have the same target).